model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/RealmCacheManager.java

```diff
@@ -109,12 +114,30 @@ public void clientRemoval(String realmId, String clientUUID, String clientId, Se
         addInvalidations(InClientPredicate.create().client(clientUUID), invalidations);
     }
 
-
     @Override
     protected void addInvalidationsFromEvent(InvalidationEvent event, Set<String> invalidations) {
         invalidations.add(event.getId());
 
         ((RealmCacheInvalidationEvent) event).addInvalidations(this, invalidations);
     }
 
+    /**
+     * Compute a cached realm and ensure that this happens only once with the current Keycloak instance.
+     * Use this to avoid concurrent preparations of a realm in parallel threads. This helps to break the load on
+     * a stampede after a server has started, were a lot of requests come in for the same realm that hasn't been cached yet.
+     * Instead of each request loading the realm in parallel, this lets the first request load the realm, and all
+     * other requests will use the cached realm, which is much more efficient.
+     */
+    public RealmAdapter computeSerialized(String id, KeycloakSession session, BiFunction<String, KeycloakSession, RealmAdapter> compute) {
+        // this locking is only to ensure that if there is a computation for the same id in the "synchronized" block below,
+        // it will have the same object instance to lock the current execution until the other is finished.
+        Object lock = cacheInteractions.computeIfAbsent(id, s -> id);
+        try {
+            synchronized (lock) {
+                return compute.apply(id, session);
+            }
+        } finally {
+            cacheInteractions.remove(lock);
+        }
```

Comment:
It does not ensure that only a single thread is computing in the same id. For example, if you have the following state
```
T1: in cacheInteractions.remove(lock);
T2: in cacheInteractions.computeIfAbsent(id, s -> id)
T3: in cacheInteractions.computeIfAbsent(id, s -> id)
```
you can have the following outcome
```
T2 returns existing `lock` instance (let's called `I1`)
T1 removes `lock` instance `I1`
T3 puts a new `lock` instance, `I2`
```
T2 and T3 will synchronize in different `lock` instances.

Reply:
@pruivo - this is true, still the behavior is sufficient to solve the cache stampede IMHO:

* When T1 removes the lock, the cache will already be filled with the realm entry
* When T2 and T3 then enter, they will get the instance from the cache, which is fine. 

The alternative would be to not remove the entry from the CHM. This would lead to a CHM with the size of all realms in Keycloak in memory, which I would like to avoid. WDYT?

Reply:
if T2 and T3 find it in the cache, even if they run in parallel, just leave it as is.

Reply:
Another thought: It might be that the cache is invalidated between in between and then I could imagine that there might parallel execution in the area of 1-2 being parallel, still the most of the other 100+ requests will be queued. 

If keeping all the IDs of the realms in memory, and it would make the design simpler, I'd be ok to make it simpler.

Reply:
I'm not sure we should use an unbounded CHM for the reasons you mentioned. A bounded map, perhaps adjustable by some (SPI) configuration option and with a good default should help.

But still, I think we should think about a more generic mechanism to also cover entity models other than the realm. And perhaps different strategies for waiting for the lock (synchronization, loop with attempts, heuristic yelds, etc). Sometimes looping a counter and parking is better than using monitors.

Reply:
IMHO it is common for multiple requests for the same realm coming in. It is less common for concurrent requests for the same client or user coming in. Therefore I'm focusing on realms here.

Not sure about not using monitors and different waiting strategies, have to think about that.

Reply:
During authentication, clients are acting on behalf of the user, and depending on the use case you end up resolving the same client across different threads too. My point is that solving for a realm will help but you still have the bottleneck somewhere else.

There are deployments where there is a handful of clients within the same realm where a set of one or two are on the user-critical path.